### ISSUE TITLE
Question regarding tf

### ISSUE THREAD

**Issue by fferreres** [dennybritz/reinforcement-learning#28]
Is there a better way to ask a question that issues? If so I won't ask again here, but if it's ok I may ask from time to time. 

1) I am reading through the DQN code, and I am not sure to understand what this part computes or outputs, other than (generally) knowing that we don't care about actions not taken, and we won't affect their classes gradients.
 
```
# Get the predictions for the chosen actions only
gather_indices = tf.range(batch_size) * tf.shape(self.predictions)[1] + self.actions_pl
self.action_predictions = tf.gather(tf.reshape(self.predictions, [-1]), gather_indices)
```

2) I am trying to understand how the backprop is done over the q(s, a) pairs (outputs of the net), and I follow the code again and again and can't find where it does step 3 and 4 like described below. In particular, each prediction will have an error in the action that is exactly what is calculated, and 0 otherwise, but it requires a vector cross all actions (ie. outputs) to let the optimizer know that the gradient only runs across one particular output. 

> Formulated in this way, the update algorithm becomes:
>
> 1) Experience a st,at,rt,st+1 transition in the environment.
> 2) Forward st+1 to evaluate the (fixed) target y=maxafθ(st+1). This quantity is interpreted to be maxaQ(st+1).
> 3) Forward fθ(st) and apply a simple L2 regression loss on the dimension at of the output, to be y. **This gradient will have a very simple form where the predicted value is simply subtracted from y. The other output dimensions have zero gradient.**
> 4) Backpropagate the gradient and perform a parameter update.
Karpathy:
http://cs.stanford.edu/people/karpathy/reinforcejs/puckworld.html

In the current code it appears it calculates an average loss given the actions for each replay and the resulting errors, calculates an average error across that play, and then updates the entire outputs based on that general error. Here's another example DQN doing the same (fixing error only to the output in the selected action):

http://neuro.cs.ut.ee/demystifying-deep-reinforcement-learning/

What am I reading wrong? Any ideas? Is the current code only actually back propagating the error only in the actions that got selected, and not the others (ie. gradient 0 on the other actions)

**Comment by fferreres** [comment]
I think I understand now. It flattens the tensor but action indexes range from (0, NUM_ACTIONS * number_batches), so it does all that above to find the index in the flattened predict tensor, and the rest is just using those indexes to get the q values of each state-action.

**Comment by dennybritz** [comment]
1. Yep, I think you got it.

2. I'm not sure if I understand the question, can you try to re-phrase it. As you mention in 1), the error is only calculated for the actions that are selected, so the gradient  for the other actions will be 0.

**Comment by fferreres** [comment]
Denny, I was royally confused as to how the graph evolved (I thought it stopped at the action outputs in the fc layer).

I hadn't noticed what tf.gather really does, and the graph will flow through it, and the backdrop starts from it and connect backwards to everything else. Naively, my question was related to this StackOverflow [question](http://stackoverflow.com/questions/36724357/in-tensorflow-can-i-use-tf-gather-for-partial-connection) 

Thanks for the quick reply. Very helpful!
